pkg/proxy/config/api.go

```diff
@@ -21,55 +21,136 @@ import (
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/kubernetes/pkg/api"
+
+	"github.com/golang/glog"
 )
 
 // NewSourceAPI creates config source that watches for changes to the services and endpoints.
 func NewSourceAPI(c cache.Getter, period time.Duration, servicesChan chan<- ServiceUpdate, endpointsChan chan<- EndpointsUpdate) {
 	servicesLW := cache.NewListWatchFromClient(c, "services", metav1.NamespaceAll, fields.Everything())
-	cache.NewReflector(servicesLW, &api.Service{}, NewServiceStore(nil, servicesChan), period).Run()
+	go NewServiceController(servicesLW, period, servicesChan).Run(wait.NeverStop)
 
 	endpointsLW := cache.NewListWatchFromClient(c, "endpoints", metav1.NamespaceAll, fields.Everything())
-	cache.NewReflector(endpointsLW, &api.Endpoints{}, NewEndpointsStore(nil, endpointsChan), period).Run()
+	go NewEndpointsController(endpointsLW, period, endpointsChan).Run(wait.NeverStop)
 }
 
-// NewServiceStore creates an undelta store that expands updates to the store into
-// ServiceUpdate events on the channel. If no store is passed, a default store will
-// be initialized. Allows reuse of a cache store across multiple components.
-func NewServiceStore(store cache.Store, ch chan<- ServiceUpdate) cache.Store {
-	fn := func(objs []interface{}) {
-		var services []api.Service
-		for _, o := range objs {
-			services = append(services, *(o.(*api.Service)))
+func sendAddService(servicesChan chan<- ServiceUpdate) func(obj interface{}) {
+	return func(obj interface{}) {
+		service, ok := obj.(*api.Service)
+		if !ok {
+			glog.Errorf("cannot convert to *api.Pod: %v", obj)
+			return
 		}
-		ch <- ServiceUpdate{Op: SET, Services: services}
+		servicesChan <- ServiceUpdate{Op: ADD, Service: service}
 	}
-	if store == nil {
-		store = cache.NewStore(cache.MetaNamespaceKeyFunc)
+}
+
+func sendUpdateService(servicesChan chan<- ServiceUpdate) func(oldObj, newObj interface{}) {
+	return func(_, newObj interface{}) {
+		service, ok := newObj.(*api.Service)
+		if !ok {
+			glog.Errorf("cannot convert to *api.Pod: %v", newObj)
+			return
+		}
+		servicesChan <- ServiceUpdate{Op: ADD, Service: service}
```

Comment:
So in the very beginning, ADD and SET were handled differently (to convey the difference).  If nothing in the subsequent steps cares about the difference (they probably shouldn't), I'd be happy having this be SET (or ADD) in all cases.